src/versionbits.cpp

```diff
@@ -57,18 +57,12 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
 
         switch (state) {
             case ThresholdState::DEFINED: {
-                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
-                    stateNext = ThresholdState::FAILED;
```

Comment:
Yes this needs coverage. I was unable to coax any other phony successful tests by modifying relevant changes in the code.

Reply:
Good catch! I think this oughta do it:

```diff
diff --git a/src/test/versionbits_tests.cpp b/src/test/versionbits_tests.cpp
index 5f53da8aff..57d9f1bedf 100644
--- a/src/test/versionbits_tests.cpp
+++ b/src/test/versionbits_tests.cpp
@@ -200,6 +200,15 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
                            .Mine(3000, TestTime(20000), 0).TestFailed().TestStateSinceHeight(3000) // 50 old blocks (so 899 out of the past 1000)
                            .Mine(4000, TestTime(20010), 0x100).TestFailed().TestStateSinceHeight(3000)
 
+        // DEFINED -> STARTED after timeout reached -> FAILED
+                           .Reset().TestDefined().TestStateSinceHeight(0)
+                           .Mine(1, TestTime(1), 0).TestDefined().TestStateSinceHeight(0)
+                           .Mine(1000, TestTime(10000) - 1, 0x100).TestDefined().TestStateSinceHeight(0) // One second more and it would be started
+                           .Mine(1999, TestTime(20000), 0).TestDefined().TestStateSinceHeight(0) // Still not started
+                           .Mine(2000, TestTime(20000), 0).TestStarted().TestStateSinceHeight(2000) // Skip straight to the timeout MTP. We guarantee one period of started
+                           .Mine(2999, TestTime(30000), 0).TestStarted().TestStateSinceHeight(2000) // 999 new blocks
+                           .Mine(3000, TestTime(30000), 0).TestFailed().TestStateSinceHeight(3000) // ... and fail
+
         // DEFINED -> STARTED -> LOCKEDIN after timeout reached -> ACTIVE
                            .Reset().TestDefined().TestStateSinceHeight(0)
                            .Mine(1, TestTime(1), 0).TestDefined().TestStateSinceHeight(0)
```

Reply:
One way to manually run the fuzz tests without running fuzzing is something like this:

```sh
ls ../../../qa-assets/fuzz_seed_corpus/versionbits/* --sort=s -r | 
while read a; do 
  if ! FUZZ=versionbits test/fuzz/fuzz < $a; then
    echo $a;
    echo 'base64:' $(base64 < $a);
    break;
  fi;
done
```

(that runs the versionbits fuzzer using the qa-assets seed corpus, in order of smallest seed to largest, and prints out the filename of the first failing seed; assumes you're in src/ already)

When I run that with the DEFINED condition changed to `if (GetMedianTimePast >= nTimeTimeout) stateNext = FAILED; else if (GetMedianTimePast >= nTimeStart) stateNext = STARTED;`, I get:

```
fuzz: test/fuzz/versionbits.cpp:322: void (anonymous namespace)::versionbits_fuzz_target(FuzzBufferType): Assertion `exp_state == ThresholdState::FAILED' failed.
../../../qa-assets/fuzz_seed_corpus/versionbits/ed29b8233d29abb3e0398be2221b9bdb8623daf6
base64: /5SUlL29vb29va0Axr3y
```

If you don't want to download the [qa-assets repo](https://github.com/bitcoin-core/qa-assets/), you should be able to duplicate that specific error with `echo '/5SUlL29vb29va0Axr3y' | base64 -d | FUZZ=versionbits test/fuzz/fuzz`

Reply:
Oh, I later setup fuzz testing but I didn't come back to check this.  I've done that just now using libFuzz and I can confirm it quickly finds this problem.

Reply:
I adapted the old `DEFINED -> FAILED` test case to become `DEFINED -> STARTED after timeout reached -> FAILED` rather than taking the case above.